migrations/src/main/scala/za/co/absa/enceladus/migrations/framework/Migrator.scala

```diff
@@ -0,0 +1,207 @@
+        db.setVersion(i)
+      }
+    }
+  }
+
+  /**
+    * Get the list of collection names valid for a particular version of the database.
+    *
+    * @param dbVersion A version number of the database.
+    */
+  def getCollectionNames(dbVersion: Int): List[String] = {
+    val collectionMigrations = getCollectionMigrations
+    var collections: List[String] = Nil
+    breakable {
+      collectionMigrations.foreach(m => {
+        if (m.targetVersion > dbVersion) {
+          break
+        } else {
+          collections = m.applyCollectionChanges(collections)
+        }
+      })
+    }
+    collections
+  }
+
+
+  /**
+    * Validates migrations for self consistency.
+    *
+    * - Target version numbers should be consequent
+    * - Collections should be added, removed or renamed consistently
+    * - Migrations should refer only to collections that do exist
+    * - Validate that migration path exists from version 0 to `targetDbVersion`
+    *
+    * @param targetDbVersion A version of the database that should be reachable by applying migrations
+    */
+  def validate(targetDbVersion: Int): Unit = {
+    validateVersionNumbersConsequent()
+    validateCollectionManipulationConsistency(targetDbVersion)
+    validateTargetVersion(targetDbVersion)
+  }
+
+  /**
+    * Validates that a list of migrations have target database versions in ascending order and there are no gaps
+    * between version numbers.
+    */
+  private def validateVersionNumbersConsequent(): Unit = {
+    var i = -1
+    migrations.foreach(m => {
+      val v = m.targetVersion
+      if (v < 0) {
+        throw new IllegalStateException(s"A negative ($v) target version is encountered in a migration spec.")
+      }
+      if (v - i > 1) {
+        throw new IllegalStateException(s"The list of migrations jumps from version $i to $v.")
+      }
+      if (v < i) {
+        throw new IllegalStateException(s"The the migrations for version $i and version $v are out of order.")
+      }
+      if (v == i) {
+        throw new IllegalStateException(s"Found 2 migrations for the same target version $i.")
+      }
+      i = v
+    })
+  }
+
+  /**
+    * Validates that a list of migrations manipulate collections consistently, e.g.
+    *  - A migration for version 0 should provide the initial list of collections
+    *  - Same collections should not be added and removed inside a single migration
+    *  - Only existing collections can be removed or renamed
+    *  - Collections should not be added twice
+    */
+  private def validateCollectionManipulationConsistency(targetDbVersion: Int): Unit = {
+    var currentVersionCollections = getCollectionNames(0)
+
+    if (currentVersionCollections.isEmpty) {
+      throw new IllegalStateException("No collection names are registered for db version 0.")
+    }
+
+    for (i <- 1 to targetDbVersion) {
+      val migrationsToExecute = migrations.filter(m => m.targetVersion == i)
```

Comment:
if we already tested that the versions do go in order, is it still needed to have such a filter instead of getting it by index? You could then drop the `foreach` on lines `145` and `146`